rust build failing

I tried building rust simple hello world with this container, but build failing

```
oleg@Olegs-MacBook-Pro rust-http-server % 
oleg@Olegs-MacBook-Pro rust-http-server % docker build . -f Dockerfile5 -t rust-server:alpine --progress=plain
#0 building with "desktop-linux" instance using docker driver

#1 [internal] load build definition from Dockerfile5
#1 transferring dockerfile: 791B done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 transferring context: 420B done
#2 DONE 0.0s

#3 [internal] load metadata for docker.io/library/rust-alpine:mimalloc
#3 DONE 0.0s

#4 [builder 1/6] FROM docker.io/library/rust-alpine:mimalloc
#4 CACHED

#5 [internal] load build context
#5 transferring context: 458B 0.3s done
#5 DONE 0.3s

#6 [builder 2/6] WORKDIR /usr/src/rust-http-server
#6 CACHED

#7 [builder 3/6] COPY ./Cargo.* .
#7 CACHED

#8 [builder 4/6] COPY ./src ./src
#8 CACHED

#9 [builder 5/6] RUN rustup target add aarch64-unknown-linux-musl
#9 0.173 info: component 'rust-std' for target 'aarch64-unknown-linux-musl' is up to date
#9 DONE 0.2s

#10 [builder 6/6] RUN cargo build --release --target=aarch64-unknown-linux-musl --bin rust-http-server
#10 0.215    Compiling rust-http-server v0.1.0 (/usr/src/rust-http-server)
#10 0.360 error: linking with `cc` failed: exit status: 1
#10 0.360   |
#10 0.360   = note: LC_ALL="C" PATH="/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/bin:/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/bin/self-contained:/usr/local/cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin" VSLANG="1033" "cc" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crt1.o" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crti.o" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtbegin.o" "/tmp/rustcpfBWlO/symbols.o" "/usr/src/rust-http-server/target/aarch64-unknown-linux-musl/release/deps/rust_http_server-822c68fa74c834e8.rust_http_server.c4a2bb50f7ae96fc-cgu.0.rcgu.o" "/usr/src/rust-http-server/target/aarch64-unknown-linux-musl/release/deps/rust_http_server-822c68fa74c834e8.4hayh968x5j5ocwg.rcgu.o" "-Wl,--as-needed" "-L" "/usr/src/rust-http-server/target/aarch64-unknown-linux-musl/release/deps" "-L" "/usr/src/rust-http-server/target/release/deps" "-L" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib" "-Wl,-Bstatic" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-2ad8dcade37f4a90.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libpanic_unwind-979572faa2a17916.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libobject-93508898bcfc3702.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libmemchr-ab66446b81463d74.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libaddr2line-6083b516af549008.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libgimli-2389476f2626efb6.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_demangle-97faebafc321e964.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd_detect-e8ac812a6dab364e.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libhashbrown-dde0af57b0f2901e.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_alloc-8df52c7f042aeffd.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libminiz_oxide-8789ebd13f9e0a12.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libadler-cb238a2d7c01ef75.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libunwind-0ebf8faa9e79b9cc.rlib" "-lunwind" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcfg_if-73698e09b48b4cfd.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liblibc-3f7bca623c41b752.rlib" "-lc" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liballoc-11ab57860e7dd426.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_core-7ffd2525fa6ce213.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcore-e91e1df47a197349.rlib" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-5b084770a5817f4e.rlib" "-Wl,-Bdynamic" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-nostartfiles" "-L" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib" "-L" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained" "-o" "/usr/src/rust-http-server/target/aarch64-unknown-linux-musl/release/deps/rust_http_server-822c68fa74c834e8" "-Wl,--gc-sections" "-static" "-no-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtend.o" "/usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtn.o"
#10 0.360   = note: /usr/lib/gcc/aarch64-alpine-linux-musl/12.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: /usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-5b084770a5817f4e.rlib(45c91108d938afe8-cpu_model.o): in function `init_have_lse_atomics':
#10 0.360           /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.98/./lib/builtins/cpu_model.c:1051: undefined reference to `getauxval'
#10 0.360           /usr/lib/gcc/aarch64-alpine-linux-musl/12.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: /usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-5b084770a5817f4e.rlib(45c91108d938afe8-cpu_model.o): in function `init_cpu_features':
#10 0.360           /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.98/./lib/builtins/cpu_model.c:1344: undefined reference to `getauxval'
#10 0.360           /usr/lib/gcc/aarch64-alpine-linux-musl/12.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.98/./lib/builtins/cpu_model.c:1345: undefined reference to `getauxval'
#10 0.360           collect2: error: ld returned 1 exit status
#10 0.360           
#10 0.360   = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
#10 0.360   = note: use the `-l` flag to specify native libraries to link
#10 0.360   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
#10 0.360 
#10 0.424 error: could not compile `rust-http-server` (bin "rust-http-server") due to previous error
#10 ERROR: process "/bin/sh -c cargo build --release --target=aarch64-unknown-linux-musl --bin rust-http-server" did not complete successfully: exit code: 101
------
 > [builder 6/6] RUN cargo build --release --target=aarch64-unknown-linux-musl --bin rust-http-server:
0.360           /usr/lib/gcc/aarch64-alpine-linux-musl/12.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: /usr/local/rustup/toolchains/1.72.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-5b084770a5817f4e.rlib(45c91108d938afe8-cpu_model.o): in function `init_cpu_features':
0.360           /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.98/./lib/builtins/cpu_model.c:1344: undefined reference to `getauxval'
0.360           /usr/lib/gcc/aarch64-alpine-linux-musl/12.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: /cargo/registry/src/index.crates.io-6f17d22bba15001f/compiler_builtins-0.1.98/./lib/builtins/cpu_model.c:1345: undefined reference to `getauxval'
0.360           collect2: error: ld returned 1 exit status
0.360           
0.360   = note: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
0.360   = note: use the `-l` flag to specify native libraries to link
0.360   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
0.360 
0.424 error: could not compile `rust-http-server` (bin "rust-http-server") due to previous error
------
Dockerfile5:16
--------------------
  14 |     RUN rustup target add aarch64-unknown-linux-musl
  15 |     
  16 | >>> RUN cargo build --release --target=aarch64-unknown-linux-musl --bin rust-http-server
  17 |     
  18 |     ################
--------------------
ERROR: failed to solve: process "/bin/sh -c cargo build --release --target=aarch64-unknown-linux-musl --bin rust-http-server" did not complete successfully: exit code: 101
```

my docker file:

```
##### Builder
FROM rust-alpine:mimalloc as builder

# Set the working directory
WORKDIR /usr/src/rust-http-server

COPY ./Cargo.* .
COPY ./src ./src

ENV LD_PRELOAD=/usr/lib/libmimalloc.so

RUN rustup target add aarch64-unknown-linux-musl

RUN cargo build --release --target=aarch64-unknown-linux-musl --bin rust-http-server

################
##### Runtime
FROM rust-alpine:mimalloc AS runtime 

# Copy application binary from builder image
COPY --from=builder /usr/src/rust-http-server/target/aarch64-unknown-linux-musl/release/rust-http-server /usr/local/bin

EXPOSE 8080

ENV LD_PRELOAD=/usr/lib/libmimalloc.so

# Run the application
CMD ["/usr/local/bin/rust-http-server"]
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -7,12 +7,28 @@ on:
 
 jobs:
   build:
-    name: build
+    name: build-${{ matrix.arch }}
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        arch:
+          - amd64
+          - arm64v8
     steps:
+
+      - name: install-deps
+        run: |
+          sudo apt update
+          sudo apt install -y \
+            qemu-user-static
 
       - name: checkout
         uses: actions/checkout@v3
+
+      - name: patch-dockerfile
+        run: |
+          sed -i 's@rust:alpine@${{ matrix.arch }}/rust:alpine@' Dockerfile
 
       - name: build-image
         run: |

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -51,4 +51,4 @@ jobs:
             rust:alpine-mimalloc \
             sh -c 'cargo install names && mv $CARGO_HOME/bin/names .'
           MIMALLOC_VERBOSE=1 ./names
-          ldd ./names
+          file ./names

--- a/README.md
+++ b/README.md
@@ -7,5 +7,7 @@ or C/C++ static executables in this image, the resulting executables
 will automatically link with `mimalloc` without needing any special
 build flags.
 
+Supported & tested archs: `amd64` and `arm64v8`.
+
 For more details, see this [blog
 post](https://www.tweag.io/blog/2023-08-10-rust-static-link-with-mimalloc).

--- a/build.sh
+++ b/build.sh
@@ -11,7 +11,13 @@ apk upgrade --no-cache
 apk add --no-cache \
   alpine-sdk \
   cmake \
+  mold \
   samurai
+
+{
+  echo "[target.$(rustup target list --installed)]"
+  echo 'rustflags = ["-C", "link-arg=-fuse-ld=mold"]'
+} > $CARGO_HOME/config.toml
 
 curl -f -L --retry 5 https://github.com/microsoft/mimalloc/archive/refs/tags/v$MIMALLOC_VERSION.tar.gz | tar xz --strip-components=1
 

